msrest/polling/poller.py

```diff
@@ -0,0 +1,214 @@
+        try:
+            self._client = client if isinstance(client, ServiceClient) else client._client
+        except AttributeError:
+            raise ValueError("Poller client parameter must be a low-level msrest Service Client or a SDK client.")
+        self._response = initial_response.response if isinstance(initial_response, ClientRawResponse) else initial_response
+        self._callbacks = []
+        # polling_method attribute is public (to be accessed from the callback)
+        self._polling_method = polling_method
+
+        if isinstance(deserialization_callback, type) and issubclass(deserialization_callback, Model):
+            deserialization_callback = deserialization_callback.deserialize
+
+        # Might raise a CloudError
+        self._polling_method.initialize(self._client, self._response, deserialization_callback)
+
+        # Prepare thread execution
+        self._thread = None
+        self._done = None
+        self._exception = None
+        if not self._polling_method.finished():
+            self._done = threading.Event()
+            self._thread = threading.Thread(
+                target=self._start,
+                name="LROPoller({})".format(uuid.uuid4()))
+            self._thread.daemon = True
+            self._thread.start()
+
+    def _start(self):
+        """Start the long running operation.
+        On completion, runs any callbacks.
+
+        :param callable update_cmd: The API reuqest to check the status of
+         the operation.
+        """
+        try:
+            self._polling_method.run()
+        except Exception as err:
+            self._exception = err
+
+        finally:
+            self._done.set()
+
+        callbacks, self._callbacks = self._callbacks, []
+        while callbacks:
+            for call in callbacks:
+                call(self._polling_method)
+            callbacks, self._callbacks = self._callbacks, []
+
+    def status(self):
+        """Returns the current status string.
+
+        :returns: The current status string
+        :rtype: str
+        """
+        return self._polling_method.status()
+
+    def result(self, timeout=None):
+        """Return the result of the long running operation, or
+        the result available after the specified timeout.
+
+        :returns: The deserialized resource of the long running operation,
+         if one is available.
+        :raises CloudError: Server problem with the query.
+        """
+        self.wait(timeout)
+        return self._polling_method.resource()
+
+    def wait(self, timeout=None):
+        """Wait on the long running operation for a specified length
+        of time.
+
+        :param int timeout: Perion of time to wait for the long running
+         operation to complete.
+        :raises CloudError: Server problem with the query.
+        """
+        if self._thread is None:
+            return
+        self._thread.join(timeout=timeout)
+        try:
+            raise self._exception
```

Comment:
This exception could be raised to the user before the thread has finished processing if the user has set a timeout - is this expected behaviour?

Reply:
I'd yes, as soon as we have an exception let's raise it

Reply:
Yes I agree we should raise the error asap, though I wonder if the user will realize that even though an exception has been raised, self._done() may not have been set yet and any callbacks will not yet have been run. Maybe this expected/standard in which case all good - though perhaps we could add a comment to the docstring stating as such?

Reply:
Yes, let's add a comment that callbacks are executed only if the poller is done. Meaning ant timeout prevents the executions of callbacks

Reply:
But the timeout wont prevent the execution of callbacks will it? Wont the thread continue?

Reply:
Hum....... that's a good point.

Reply:
Ok, so you confused me with your first comment :p
If `_exception` is set, this is finished anyway. So yes, it was the expected behvavior. I will revert my commit with force callbacks